Way to configure TLS application protocols

jitsni opened the pull request.

jitsni committed:
Proving a way to configure application protocols for TLS alpn extension

jitsni committed:
removing the commented argLine

jitsni committed:
SSLEngine#setHandshakeApplicationProtocolSelector() to select application protocol
during handshake on server side. It selects the application protocol of the matched
server.

jitsni committed:
Writing application protocol in the BEGIN frame extension on the reply stream

jitsni committed:
Merge remote-tracking branch 'upstream/develop' into alpn

Conflicts:
	pom.xml
	src/main/java/org/reaktivity/nukleus/tls/internal/stream/ClientStreamFactory.java
	src/main/java/org/reaktivity/nukleus/tls/internal/stream/ServerStreamFactory.java

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ClientStreamFactory.java:
```diff
@@ -250,13 +260,15 @@ private RouteFW wrapRoute(
 
         private ClientAcceptStream(
             String tlsHostname,
+            String applicationProtocol,
```
`applicationProtocol` -> `tlsApplicationProtocol`

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ClientStreamFactory.java:
```diff
@@ -340,7 +352,7 @@ private void handleBegin(
-                String[] applicationProtocols = config.clientApplicationProtocols();
+                String[] applicationProtocols = new String[] { applicationProtocol };
                 if (applicationProtocols.length > 0)
                 {
```
The code logic has changed here. What is the expected behavior when `tlsApplicationProtocol` is `null`?

Also need to remove the unnecessary `try-catch` around `tlsParameters.setApplicationProtocols(...)` given that the TLS nukleus requires Java 9 in order to run anyway.

jitsni replied:
try-catch is not there anymore.

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ServerStreamFactory.java:
```diff
@@ -352,6 +336,44 @@ private void handleBegin(
             }
         }
 
+        private String selectApplicationProtocol(SSLEngine tlsEngine, List<String> clientProtocols)
```
Method parameters on next line, surprised this wasn't caught by checkstyle.

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ServerStreamFactory.java:
```diff
@@ -640,7 +662,18 @@ private void handleFinished()
+                if (applicationProtocol == null || applicationProtocol.isEmpty())
+                {
+                    applicationProtocol = "";
+                }
```
This code implies that the `TlsBeginEx` does not always contain the application protocol that was actually negotiated  during the TLS handshake, which doesn't seem right.

Special case for empty string doesn't look right either.

What breaks if lines `666-673` are removed?

Suggest also moving line `665` to `658` and renaming to `tlsApplicationProtocol` alongside `tlsHostname`.

jitsni replied:
Not applicable anymore

jitsni committed:
Minor reformatting
s/applicationProtocol/tlsApplicationProtocol

jitsni committed:
TLS application protocol can be set as null (now flyweights support null strings)

jitsni committed:
renaming applicationProtocol/tlsApplicationProtocol

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ClientStreamFactory.java:
```diff
@@ -174,12 +172,12 @@ private MessageConsumer newAcceptStream(
                     applicationName.equals(route.source().asString()) &&
                     (tlsHostname == null || Objects.equals(tlsHostname, hostname)) &&
-                    (applicationProtocol == null || applicationProtocol.equals(routeEx.applicationProtocol().asString()));
+                    (tlsApplicationProtocol == null || tlsApplicationProtocol.equals(routeEx.applicationProtocol().asString()));
```
Hoist `routeEx.applicationProtocol().asString()` as `applicationProtocol` above, alongside `hostname`.

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ClientStreamFactory.java:
```diff
@@ -352,17 +350,10 @@ private void handleBegin(
 
-                String[] applicationProtocols = new String[] { applicationProtocol };
-                if (applicationProtocols.length > 0)
+                if (tlsApplicationProtocol != null && !tlsApplicationProtocol.isEmpty())
```
We should only test for `!= null` here, since empty string is a valid protocol value.

jitsni replied:
SSLParameters#setApplicationProtocols() doesn't allow empty string for protocol

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ClientStreamFactory.java:
```diff
@@ -613,12 +604,12 @@ private MessageConsumer doBeginApplicationReply(
         {
             final String applicationReplyName = applicationName;
             final String peerHost = tlsEngine.getPeerHost();
```
`peerHost` -> `tlsPeerHost` if using `tlsApplicationProtocol` naming convention for `applicationProtocol`.

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ServerStreamFactory.java:
```diff
@@ -660,20 +663,15 @@ private void handleFinished()
-                }
-                if (applicationProtocol == null || applicationProtocol.isEmpty())
+                String tlsApplicationProtocol = tlsEngine.getApplicationProtocol();
+                if (tlsApplicationProtocol == null || tlsApplicationProtocol.isEmpty())
```
Only test for `!= null` since empty string is a valid value for application protocol.

jitsni replied:
SSLEngine#getApplicationProtocol() returns "". From javadoc,

"null if it has not yet been determined if application protocols might be used for this connection, an empty String if application protocols values will not be used, or a non-empty application protocol String if a value was successfully negotiated."

jfallows replied:
good clarification - thanks, let's revert `""` back to `null` to it appears "unset" for routing purposes.

jfallows commented on src/main/java/org/reaktivity/nukleus/tls/internal/stream/ServerStreamFactory.java:
```diff
@@ -660,20 +663,15 @@ private void handleFinished()
-                    applicationProtocol = tlsRouteEx.applicationProtocol().asString();
-                }
-                if (applicationProtocol == null || applicationProtocol.isEmpty())
+                String tlsApplicationProtocol = tlsEngine.getApplicationProtocol();
```
Move this to 661 alongside `tlsHostname`.

jitsni committed:
minor reformatting and renaming changes

jitsni committed:
Once handshake is finished, route selection should include application protocol
logic

jfallows committed:
Update nukleus-maven-plugin version

jitsni committed:
Opting out of ALPN negotiation when the specific route is not matched,
but only default route is matched. It behaves as if there is no ALPN
negotation

jitsni committed:
Adding additional checks once the handshake is complete to make
sure a correct route can still be used. The following cases
would succeed otherwise a RESET the connection
* no ALPN negotiation but there is a default route
* negotiated protocol from ALPN matches with our route

jitsni committed:
Merge branch 'alpn' of github.com:jitsni/nukleus-tls.java into alpn

jfallows commented on src/test/java/org/reaktivity/nukleus/tls/internal/streams/ServerIT.java:
```diff
@@ -70,7 +80,97 @@ public void shouldEstablishConnection() throws Exception
+     * client omits ALPN
+     * negotiates successfully without ALPN, but protocol != protocol2 => RESET transport
+     */
+    @Ignore("There no way to detect connection abort in k3po")
```
File a ticket for this if it doesn't already exist, and refer to the ticket from this `@Ignore` annotation.

jitsni committed:
Linking k3po issue with the ignored tests

jfallows committed:
Update k3po version

jfallows committed:
Update nukleus-tls.spec dependency version
